core/primitives/src/receipt.rs

```diff
@@ -90,6 +91,71 @@ pub enum Receipt {
     V1(ReceiptV1),
 }
 
+/// A receipt that is stored in the state with added metadata. A receipt may be
+/// stored in the state as a delayed receipt, buffered receipt or a promise
+/// yield receipt. The metadata contains additional information about receipt
+///
+/// Please note that the StateStoredReceipt implements custom serialization and
+/// deserialization. Please see the comment on [ReceiptOrStateStoredReceipt]
+/// for more details.
+///
+/// This struct is versioned so that it can be enhanced in the future.
+pub enum StateStoredReceipt {
+    V0(StateStoredReceiptV0),
+}
+
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct StateStoredReceiptV0 {
+    /// The receipt.
+    pub receipt: Receipt,
+    pub metadata: StateStoredReceiptMetadata,
+}
```

Comment:
It's a bit sad that the receipt and metadata are stored together.

Witness size limits currently have a problem with reading receipts from trie queues. We can read the receipt, but we don't know how large it is until we read it. Reading a receipt of unkown size might suddenly increase witness size. It'd be amazing to know the size before we decide whether to read it from state or not.
I think it would be nicer to store the metadata separately. It'd avoid the serialization hackery and make things easier for witness size limit.

Reply:
> we don't know how large it is until we read it

Now that I think about it, it might be possible to implement something like this. Read the trie node, but not the value, something like `has_key`.

Reply:
Thanks, that's a good idea! 
One way to go about it would be to have a parallel queue / buffers for the existing queue and buffers that stores only the metadata. Would that solve the problem? It would double the number of trie accesses for ops on those queue and buffers though. Did you have any other idea for this?

Reply:
> One way to go about it would be to have a parallel queue / buffers for the existing queue and buffers that stores only the metadata. Would that solve the problem? It would double the number of trie accesses for ops on those queue and buffers though. Did you have any other idea for this?

Yeah that's pretty much what I had in mind - a separate `TrieKey` that's indexed using the same indexes as the items in the original queue. Something like `TrieKey::DelayedReceiptMetadata { index: u64 }`. We would save the metadata when an item is added to the queue and remove it when it's removed. For items that were saved before the protocol change there would be no metadata, we can handle that.

Reply:
One more trie access shouldn't be that much of a problem with memtrie

Reply:
> Now that I think about it, it might be possible to implement something like this even without storing metadata separately. Read the trie node, but not the value, something like has_key.

But now I'm thinking that it might be possible to get the size without saving separate metadata, so maybe there's no point in rewriting everything. It's even more efficient to save them together.

Reply:
And for bandwidth limits C2 we're gonna have to change what's kept in the delayed queue, so the hackery will be required anyway.

Reply:
All right, I'll leave it as is for now then.